docker/pytorch/Dockerfile

```diff
@@ -185,3 +185,19 @@ RUN sudo pip${PYTHON_VERSION} install --no-cache-dir \
     cupy-`echo ${CUDA_VERSION_TAG} | sed "s/cu/cuda/g"`>=10.2.0 && \
     sudo pip${PYTHON_VERSION} install --no-cache-dir mmcv-full==${MMCV_VERSION} -f https://download.openmmlab.com/mmcv/dist/${CUDA_VERSION_TAG}/${MMCV_TORCH_VERSION}/index.html && \
     sudo pip${PYTHON_VERSION} install --no-cache-dir 'mmsegmentation>=0.22.0,<0.23'
+
+
+FROM pytorch_stage AS mofed_stage
```

Comment:
I think there's a mismatch here between what we want it to do and what it will actually do.

The way staged builds work is the specified target is actually a stop target so it will actually try running all the other stages before this stage.  It's not completely clear to me but it's documented here:
https://docs.docker.com/develop/develop-images/multistage-build/#stop-at-a-specific-build-stage

So I don't think this is going to run just `pytorch_stage` followed by `mofed_stage`, and since you're not specifying any of the arguments required for the `vision_stage` I'm guessing this will error out.

Few options to workaround:
1. Set defaults for `CUDA_VERSION_TAG`, `MMCV_VERSION`, `MMCV_TORCH_VERSION` and `PYTHON_VERSION` in the `vision_stage` in the Dockerfile or in the build matrix so the stage can at least build.  Because you're resuming `mofed_stage` from `pytorch_stage` I don't believe any of the `vision_stage` layers will be in the final image
2. Move the `mofed_stage` _before_  `vision_stage` and set defaults for the arguments
3. Build the MOFED dependencies into the `pytorch_stage` image.
4. Add another Dockerfile for the MOFED image

From a UX perspective I'm not really sure it makes sense to have separate images for single-node vs. multi-node anyways so I wouldn't be opposed to just going with option 3 and keeping the multi-stage stuff simple for now.

Options 1 and 2 are really the same problem just swapped and 4 is probably going to involve some Jenkins wrangling.

Reply:
Yeah so I noticed this too.

I think this is actually a bug with `kaniko`, the tool we use for generating Docker images in the Jenkins process. When I just run this build with `docker build` locally, I actually do only see the `pytorch_stage` and `mofed_stage` running.

Option 1 is the workaround I was quietly shooting for; if you look at my recent commits I've been adding those defaults. I don't really care if the `kaniko` build is longer than it's supposed to be - I might even submit an issue to that repo. I think there's a bug in the defaults I'm adding, but once that's resolved the build should succeed.

Option 3 is something I think we'd discussed explicitly wanting to avoid. The Mellanox drivers are substantial in size, and I think @hanlint may have mentioned that we'd prefer to avoid including them by default?

Option 4 would work but I don't feel like Jenkins wrangling right now, but if you put a gun to my head I'll do it.

Reply:
FWIW the OFED packages are adding ~1.6GB of bloat to the `mosaicml/pytorch` image so may want to just consider adding these dependencies to the base images unless there's some functional reason not to vs. creating another stage target:
`REPOSITORY                TAG                                  IMAGE ID       CREATED        SIZE
mosaicml/pytorch_vision   1.11.0_cu113-python3.9-ubuntu20.04   9f2b3a75aba9   21 hours ago   14.3GB
mosaicml/pytorch          1.11.0_cu113-python3.9-ubuntu20.04   b52fc4b3663f   21 hours ago   12.5GB
mosaicml/pytorch_mofed    jamie-temp                           5d27b1a2405e   24 hours ago   14.1GB`

Reply:
Alright well personally, I'm totally fine with adding the OFED packages to `mosaicml/pytorch`. I don't remember who it was with the objection - maybe @ravi-mosaicml ? I'll make the change, and if I get a sign off I'll merge it.

Reply:
> I think this is actually a bug with `kaniko`, the tool we use for generating Docker images in the Jenkins process. When I just run this build with `docker build` locally, I actually do only see the `pytorch_stage` and `mofed_stage` running.

Gotcha that would make more sense at least intuitively, the docs aren't really clear on this.